### ISSUE TITLE
New Android NDK support request

### ISSUE THREAD

**Issue by holmesconan** [microsoft/vcpkg#21345]
Referenced from [NDK Revision History](https://developer.android.com/ndk/downloads/revision_history), since Android NDK r11 (March 2016), clang is the strongly recommended compiler and toolchain. And the GCC deprecation warning shown from Android NDK r13b (October 2016). Then since Android NDK r18b (September 2018), GCC is removed from the toolchan and only clang is supported.

When do cross-compile with the latest Android NDK r23 for any autotools based library like libiconv, errors will raise as:

```
...
-- Configuring x86-android-dbg
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:127 (message):
    Command failed: /usr/bin/bash -c "V=1 ./../src/1.16-b0c41d3469.clean/configure  \"--enable-extra-encodings\" \"--without-libiconv-prefix\" \"--without-libintl-prefix\" \"--disable-silent-rules\" \"--verbose\" \"--disable-shared\" \"--enable-static\" \"--prefix=/home/vcpkg/installed/x86-android/debug\" \"--bindir=\\${prefix}/../tools/libiconv/debug/bin\" \"--sbindir=\\${prefix}/../tools/libiconv/debug/sbin\" \"--libdir=\\${prefix}/lib\" \"--includedir=\\${prefix}/../include\" \"--datarootdir=\\${prefix}/share/libiconv\""
    Working Directory: /home/vcpkg/buildtrees/libiconv/x86-android-dbg
    Error code: 77
    See logs for more information:
      /home/vcpkg/buildtrees/libiconv/config-x86-android-dbg-out.log
      /home/vcpkg/buildtrees/libiconv/config-x86-android-dbg-err.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_configure_make.cmake:796 (vcpkg_execute_required_process)
  ports/libiconv/portfile.cmake:29 (vcpkg_configure_make)
  scripts/ports.cmake:142 (include)


Error: Building package libiconv:x86-android failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `git pull` and `./vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  package: libiconv[core]:x86-android -> 1.16#11
  vcpkg version: 2021-11-02-af04ebf6274fd6f7a941bff4662b3955c64f6f42
  vcpkg-tool version: 897ff9372 2021-11-11 (9 hours ago)

Additionally, attach any relevant sections from the log files above.
```

And `config-x86-android-dbg-err.log` shown

```
configure: error: in `/home/vcpkg/buildtrees/libiconv/x86-android-dbg':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

And `config-x86-android-dbg-out.log` shown

```
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make sets $(MAKE)... (cached) yes
checking for gcc... gcc
checking whether the C compiler works... no
```

Obviously, the `configure` script did not find the right compiler for cross-compile.

**Proposed solution**
issue #20844 and draft PR #21222 proposed a temporary solution that forces provides `CC` and `CXX` (and other tools like strip, ar, ranlib, etc). But it just solved libraries that used autotools build system.

other minor fixes for CMake build system. Such as `bzip2:x86-android`. 

```
$ cat buildtrees/bzip2/config-x86-android-dbg-out.log
-- ANDROID_PLATFORM not set. Defaulting to minimum supported version 16.
-- Android: Targeting API '16' with architecture 'x86', ABI 'x86', and processor 'i686'
-- Android: Selected unified Clang toolchain
-- The C compiler identification is Clang 12.0.8
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/AndroidSdk/ndk/23.1.7779620/toolchains/llvm/prebuilt/linux-x86_64/bin/clang - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vcpkg/buildtrees/bzip2/x86-android-dbg
```

even though the library is build without error.

```
$ head -n 10 scripts/toolchains/android.cmake
set(ANDROID_CPP_FEATURES "rtti exceptions" CACHE STRING "")
set(CMAKE_SYSTEM_NAME Android CACHE STRING "")
set(ANDROID_TOOLCHAIN clang CACHE STRING "")
set(ANDROID_NATIVE_API_LEVEL ${CMAKE_SYSTEM_VERSION} CACHE STRING "")
set(CMAKE_ANDROID_NDK_TOOLCHAIN_VERSION clang CACHE STRING "")
```

According to the [reference](https://developer.android.com/ndk/guides/cmake), `ANDROID_NATIVE_API_LEVEL` is the alias of `ANDROID_PLATFORM`, but one needs to specify both instead of one.

**Describe alternatives you've considered**
There are many other build systems that vcpkg supported. And CMake is the primary build system now (instead of ndk-build). And there is a CMake toolchain file that prepared the build environment. So after the temporary solution, we need to reconsider the whole picture for the Android triplet.

As the reviewer of #21222 @Neumann-A said:

> Again I feel like the default android toolchain vcpkg provides should be removed and replaced with specialized toolchains for each architecture and case. There seems to be a bigger misunderstanding of vcpkg is supposed to work and what work should be done by the configure scripts.

And we need more consideration of how to pass

1. ANDROID_ABI
2. ANDROID_PLATFORM
3. ANDROID_API_LEVEL
4. etc

to most build systems. More discussion please read the review comment of #21222 .

**Comment by m-kuhn** [comment]
Thank you for opening this issue and the work in the PR. It would be nice to have a solution for android in vcpkg.
For reference, I'd also like to link previous work by @xandox #15605

**Comment by holmescn** [comment]
@m-kuhn I have read PR #15605. Although it is outdated, it provided another kind of solution. Actually, the very first solution of mine is almost the same except I directly set the ENV{CC} from the block of `if (VCPKG_TARGET_IS_ANDROID)`. I could follow the solution to make another PR. But I need to find out where the `set(ENV{CC} ${VCPKG_DETECTED_CMAKE_C_COMPILER})` should be placed?

**Comment by holmescn** [comment]
Another point to discuss, as mentioned by @Neumann-A, should we removed the default android toolchain vcpkg provides in `<VCPKG_ROOT>/toolchains/android.cmake` and use the one provided by Android NDK in `<ANDROID_NDK_HOME>/build/cmake/android.toolchain.cmake`

**Comment by holmescn** [comment]
Then the third question to be discussed:

Should we specify the ANDROID_PLATFORM/ANDROID_NATIVE_API_LEVEL in `triplets/community/*-android.cmake` files? Then anyone could use the standard `$VCPKG_OVERLAY_TRIPLETS` to customize their build?

**Comment by Neumann-A** [comment]
>  and use the one provided by Android NDK in

My whole point is
- you want to set `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` to some sort of custom android toolchain in the specialized android triplet 
- the custom android toolchain specializes on your use case and setups the the required variables which are consumed and used by `<ANDROID_NDK_HOME>/build/cmake/android.toolchain.cmake`. The NDK toolchain is included after that in the custom android toolchain (and should thus be correctly setup)
- the configure scripts in cmake have a section `if(VCPKG_TARGET_ANDROID)` which setups the minimal additional settings to make it work. Important don't do stuff here which is also required by other targets (which would be for example setting up CC and stuff; cmake should have selected the right compiler from the start; if it is not correctly setup, set it up correctly in general and not just for android.) In general only required env variables like `ANDROID_PLATFORM` (if required at all!) should be additionally setup. 
- don't take the `VCPKG_TARGET_IS_UWP` stuff as an example in the configure scripts. The if block is conceptually wrong and the setup should be done in the windows toolchain which unfortunately is not correctly setup for UWP which is also why CMake/vcpkg is not using ninja for UWP... (it relies on the fact that the generated proj files will have correct defaults. )

**Comment by holmescn** [comment]
@Neumann-A if I understood correctly, there are two points of your view:

1. customize or set ANDROID_API_LEVEL etc in custom android toolchain file through VCPKG_CHAINLOAD_TOOLCHAIN_FILE  instead of ENV{ANDROID_API_LEVEL}
2. set ENV{CC} like variables in general place instead of Android specialized block

**Comment by luncliff** [comment]
https://github.com/microsoft/vcpkg/issues/21345#issuecomment-966979955 is correct. 
I think those NDK(both toolkit path and build environment) related issue can be handled with customized `*-android.cmake` triplets.

https://github.com/luncliff/vcpkg-registry/blob/0.2146.0/triplets/arm64-android.cmake

**Comment by holmescn** [comment]
```
$ find $VCPKG_ROOT/scripts -iname "*.cmake" -exec grep "toolchains" {} \; -print
```
shows 
```
    # Load toolchains
        set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/windows.cmake")
/home/vcpkg/scripts/cmake/vcpkg_build_nmake.cmake
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/windows.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/linux.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/android.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/osx.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/ios.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/freebsd.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/openbsd.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/mingw.cmake")
/home/vcpkg/scripts/cmake/vcpkg_configure_cmake.cmake
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/windows.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/linux.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/android.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/osx.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/ios.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/freebsd.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/openbsd.cmake")
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/mingw.cmake")
/home/vcpkg/scripts/cmake/vcpkg_configure_meson.cmake
            set(VCPKG_CHAINLOAD_TOOLCHAIN_FILE "${SCRIPTS}/toolchains/mingw.cmake" PARENT_SCOPE) # Switching to MinGW toolchain for Fortran
/home/vcpkg/scripts/cmake/vcpkg_find_fortran.cmake
```

which indicated that only

1. `scripts/cmake/vcpkg_configure_meson.cmake`
2. `/home/vcpkg/scripts/cmake/vcpkg_configure_cmake.cmake`

chained the Android toolchain file. Does it mean that `vcpkg_configure_make.cmake` have to specify the compiler by itself through the detected environment by `get_cmake_vars`?

**Comment by Neumann-A** [comment]
> chained the Android toolchain file. Does it mean that vcpkg_configure_make.cmake have to specify the compiler by itself through the detected environment by get_cmake_vars?

Your CMake toolchain needs to setup the compiler. Meson then will take it via `get_cmake_vars`. The same is true for autotools (vcpkg_configure_make although that might currently be missing the general setup outside of windows).

**Comment by holmescn** [comment]
@Neumann-A is there some documentations for understanding what vcpkg was doing after run `vcpkg install <something>`?

**Comment by holmescn** [comment]
> although that might currently be missing the general setup outside of windows.

Yes, that is why is cancelled the previous PR. We need a full refractory for the `vcpkg_configure_make`. First, we need to divide the build process into three parts:

1. the most general part for both host build and cross-platform build
2. the specialized part for just host build
3. the specialized part for cross-platform build

But in the first place, I need to understand how variables go through the whole build process. I have tested that variables that are set inside `triplets/community/*-android.cmake` are not visible in `scripts/toolchains/android.cmake`, so do the `vcpkg_configure_make.cmake` scope.

**Comment by dg0yt** [comment]
> is there some documentations for understanding what vcpkg was doing after run `vcpkg install <something>`?

There is not so much documentation, but running `vcpkg --debug` will show you some entry points. Basically, vcpkg runs `scripts/ports.cmake` (in script mode - see maintainer guide!).

Loading the files from `scripts/toolchains` in script mode isn't good practice IMO. In script mode, without languages, tools etc. configured, they cannot have full effect. That's why there is the helper function which detects variables by running CMake in project mode.

`vcpkg_configure_make` already sets all relevant variables for building *for the target*, deriving the values from what a cmake project would use. However, there is a gap in configuration when it comes to building tools for the host. vcpkg expect this to be solved via host dependencies, but this needs more work.

**Comment by holmescn** [comment]
@dg0yt yes, I am trying to understand the build process and do some refactoring. I think the triplets files are good place to customize the build, perhaps pending the API level after the triplet name like `*-android23` instead of `*-android`. Since the ABI are not compatible. So I tried to pass the `ANDROID_NATIVE_API_LEVEL` and `ANDROID_PLATFORM` from triplet to the toolchain but failed. More try-and-errors should be done.

**Comment by Neumann-A** [comment]
@holmescn You can also run `vcpkg install` with `--x-cmake-args=--trace-expand` to get trace output of what vcpkg is doing regarding to its cmake scripts.

**Comment by dg0yt** [comment]
Note that the cmake variables from the triplet are seen in script mode ( `portfile.cmake`), but not during the actual project configuration (`CMakeLists.txt` and its toolchain).
However, environment variables set in the triplet may be seen during project configuration - if vcpkg is told to pass them through.

**Comment by holmescn** [comment]
```
$ cat triplets/community/x86-android.cmake
set(VCPKG_TARGET_ARCHITECTURE x86)
set(VCPKG_CRT_LINKAGE static)
set(VCPKG_LIBRARY_LINKAGE static)
set(VCPKG_CMAKE_SYSTEM_NAME Android)
```
shown that there are four variables set by the triplet.

```
$ find scripts -iname "*.cmake" -exec grep -q "VCPKG_TARGET_ARCHITECTURE" {} \; -print
scripts/buildsystems/vcpkg.cmake
scripts/cmake/vcpkg_fail_port_install.cmake
scripts/cmake/vcpkg_find_acquire_program.cmake
scripts/cmake/vcpkg_configure_cmake.cmake
scripts/cmake/vcpkg_configure_meson.cmake
scripts/cmake/vcpkg_configure_make.cmake
scripts/cmake/vcpkg_install_msbuild.cmake
scripts/cmake/vcpkg_find_fortran.cmake
scripts/templates/portfile.in.cmake
scripts/vcpkg_get_dep_info.cmake
scripts/vcpkg_get_tags.cmake
scripts/ports.cmake
scripts/toolchains/linux.cmake
scripts/toolchains/mingw.cmake
scripts/toolchains/ios.cmake
scripts/test_ports/cmake/portfile.cmake

$  find scripts -iname "*.cmake" -exec grep -q "VCPKG_CRT_LINKAGE" {} \; -print
scripts/build_info.cmake
scripts/cmake/vcpkg_check_linkage.cmake
scripts/cmake/vcpkg_fail_port_install.cmake
scripts/cmake/vcpkg_configure_cmake.cmake
scripts/cmake/vcpkg_configure_meson.cmake
scripts/cmake/vcpkg_configure_qmake.cmake
scripts/cmake/vcpkg_find_fortran.cmake
scripts/templates/portfile.in.cmake
scripts/vcpkg_get_dep_info.cmake
scripts/vcpkg_get_tags.cmake
scripts/toolchains/linux.cmake
scripts/toolchains/android.cmake
scripts/toolchains/windows.cmake
scripts/toolchains/mingw.cmake

$ find scripts -iname "*.cmake" -exec grep -q "VCPKG_CRT_LINKAGE" {} \; -print
scripts/build_info.cmake
scripts/cmake/vcpkg_check_linkage.cmake
scripts/cmake/vcpkg_fail_port_install.cmake
scripts/cmake/vcpkg_configure_cmake.cmake
scripts/cmake/vcpkg_configure_meson.cmake
scripts/cmake/vcpkg_configure_qmake.cmake
scripts/cmake/vcpkg_find_fortran.cmake
scripts/templates/portfile.in.cmake
scripts/vcpkg_get_dep_info.cmake
scripts/vcpkg_get_tags.cmake
scripts/toolchains/linux.cmake
scripts/toolchains/android.cmake
scripts/toolchains/windows.cmake
scripts/toolchains/mingw.cmake
┌[holmes☮gorgeous]-(/home/vcpkg)-[git://enhencement-new-android-ndk ✔]-
└> find scripts -iname "*.cmake" -exec grep -q "VCPKG_LIBRARY_LINKAGE" {} \; -print
scripts/build_info.cmake
scripts/cmake/vcpkg_check_linkage.cmake
scripts/cmake/vcpkg_build_msbuild.cmake
scripts/cmake/vcpkg_fail_port_install.cmake
scripts/cmake/vcpkg_install_meson.cmake
scripts/cmake/vcpkg_build_make.cmake
scripts/cmake/vcpkg_configure_cmake.cmake
scripts/cmake/vcpkg_configure_meson.cmake
scripts/cmake/vcpkg_configure_make.cmake
scripts/cmake/vcpkg_install_msbuild.cmake
scripts/cmake/vcpkg_fixup_pkgconfig.cmake
scripts/cmake/vcpkg_configure_qmake.cmake
scripts/cmake/vcpkg_find_fortran.cmake
scripts/cmake/vcpkg_copy_pdbs.cmake
scripts/boost/post-build-stubs/config.cmake
scripts/templates/portfile.in.cmake
scripts/vcpkg_get_dep_info.cmake
scripts/vcpkg_get_tags.cmake

$ find scripts -iname "*.cmake" -exec grep -q "VCPKG_CMAKE_SYSTEM_NAME" {} \; -print
scripts/cmake/vcpkg_common_definitions.cmake
scripts/cmake/vcpkg_configure_cmake.cmake
scripts/cmake/vcpkg_configure_meson.cmake
scripts/vcpkg_get_dep_info.cmake
scripts/vcpkg_get_tags.cmake
scripts/toolchains/mingw.cmake
```
shown where they are used. @dg0yt given an awesome hint.

**Comment by dg0yt** [comment]
Note that not everything in the existing scripts is what I would call best practice. The key challenge is to scale well over different targets and over time. 
Some scripts functions are slowly moved to ports (e.g. vcpkg-cmake etc.) so that they can be versioned.
Helper functions from ports are loaded in script mode via `vcpkg-cmake-config.cmake` if the using port declares a (host) dependency on the providing port.

**Comment by holmescn** [comment]
> Note that not everything in the existing scripts is what I would call best practice.  The key challenge is to scale well over different targets and over time.

Yes, that is a great challenge. That is why I just add a minor patch to fix just android and leave other platforms alone. I think the open-source community should just do the specialized patches, and the generalized architecture and structure should be designed by the vcpkg team. But the previous PR use the ENV in a bad style since Android ABIs are not fully compatible. One may eventually change the Android API level then all installed libraries will not be linked. That would be a problem to be solved in the next stage.

**Comment by holmescn** [comment]
Almost there. When run
```
$ vcpkg install --debug --x-cmake-args=--trace-expand bzip2:x86-android
```
I found this line in the log
```
[DEBUG] popen(/home/vcpkg/downloads/tools/cmake-3.21.1-linux/cmake-3.21.1-linux-x86_64/bin/cmake -DALL_FEATURES= -DCURRENT_PORT_DIR=/home/vcpkg/ports/bzip2 -D_HOST_TRIPLET=x64-linux -DFEATURES=core -DPORT=bzip2 -DVCPKG_USE_HEAD_VERSION=0 -D_VCPKG_DOWNLOAD_TOOL=BUILT_IN -D_VCPKG_EDITABLE=1 -D_VCPKG_NO_DOWNLOADS=0 --trace-expand -DCMD=BUILD -DDOWNLOADS=/home/vcpkg/downloads -DTARGET_TRIPLET=x86-android -DTARGET_TRIPLET_FILE=/home/vcpkg/triplets/community/x86-android.cmake -DVCPKG_BASE_VERSION=2021-11-15 -DVCPKG_CONCURRENCY=65 -DVCPKG_PLATFORM_TOOLSET=external -DGIT=/usr/bin/git -DVCPKG_ROOT_DIR=/home/vcpkg -DPACKAGES_DIR=/home/vcpkg/packages -DBUILDTREES_DIR=/home/vcpkg/buildtrees -D_VCPKG_INSTALLED_DIR=/home/vcpkg/installed -DDOWNLOADS=/home/vcpkg/downloads -DVCPKG_MANIFEST_INSTALL=OFF -P /home/vcpkg/scripts/ports.cmake･2>&1)
```
This seems to be the start point of the build process. The `ports.cmake` run in script mode with information collected from vcpkg command-line arguments and environment. As mentioned in the CMake documentation, the script mode just run the script file, no configure or generate step is performed. Then I read the `ports.cmake`, it just has 190 lines of code. Before line 63, all utility functions are included. So any variable set in triplet will not be accessible in those functions. Then between line 97 and 146 defined the "BUILD" command. I think it is the core part. After a bunch of checks, at line 128, the triplet file was included. Then line 142 included the `portfile.cmake` that actually call any utility functions included before line 63 and does the actual build process.

Since when the actual triplet file is included, the package dir and installed dir are made. So I wish to use like `x86-android23` to store all libraries build for Android API level 23 is not possible. I think one can add an overlay triplet named `x86-android23` to accomplish the goal. 

What is still confused is that where is the toolchain file included?

**Comment by dg0yt** [comment]
> What is still confused is that where is the toolchain file included?

In `vcpkg_configure_make`, the toolchain is used via `z_vcpkg_get_cmake_vars`, creating`VCPKG_DETECTED_...` variables.
However, I was wrong when I assumed that `vcpkg_configure_make` would generally use `VCPKG_DETECTED_CMAKE_C_COMPILER` etc. :disappointed: Unfortunately it is hard to improve the file, because it takes long in the CI, it was blocked by a script audit, and there is a new effort to move it to a port instead.

**Comment by m-kuhn** [comment]
I opened a PR against your branch which makes use of this

https://github.com/holmescn/vcpkg/pull/1

**Comment by holmescn** [comment]
@dg0yt before anything moved into port, let's try to accomplish this:

1. add `set(VCPKG_ANDROID_API_LEVEL 23)` into `triplets/community/*-android.cmake` to specify the Android API version
2. parse `$ANDROID_NDK_HOME/source.properties` to detect Android NDK version in `toolchains/android.cmake`, then set specialized `ANDROID_PLATFORM` or `ANDROID_NATIVE_API_LEVEL` which is in PR #21465 

the simple way to accomplish the first goal is using environment variables like
```
set(ENV{VCPKG_ANDROID_API_LEVEL} 23)
```
but I don't know is it acceptable @Neumann-A. Then there is a problem that old NDK perhaps not support some newer version of Android API. for example, if one used, let's say android ndk r15d, which may not accept Android API level 23.

The `z_vcpkg_get_cmake_vars` isn't the problem. We can do it as the last PR and #15605. Just use the detected C/C++ compiler and if it it clang, add CFLAGS and CXXFLAGS with `--triplet`.

**Comment by holmescn** [comment]
@m-kuhn thank you for your PR. But that branch will mark as a reference until we found the next solution.

**Comment by m-kuhn** [comment]
What is the next solution?
My PR takes care of
- the API version. By using `z_vcpkg_get_cmake_vars` this is the same as is used for other ports.
- the path for tools

**Comment by holmescn** [comment]
@m-kuhn since I have closed my PR #21222, I thought you can make a draft PR here. I think it is a good solution for `vcpkg_configure_make`.  The next solution I mean is that add some customization in triplet and use it inside most build systems not only the `vcpkg_configure_make`.

**Comment by holmescn** [comment]
In `$VCPKG_ROOT/scripts/toolchains/android.cmake`, the `$ANDROID_NATIVE_API_LEVEL` was set to `$CMAKE_SYSTEM_VERSION`. Does anyone know that where did this variable set?

**Comment by holmescn** [comment]
Hi everyone followed this issue, the new PR is made, please review it and give me any feedback.

**Comment by dg0yt** [comment]
#21831 misses a mention here.

**Comment by holmescn** [comment]
@dg0yt how can I test the PR locally?

**Comment by dg0yt** [comment]
@holmescn At the top of the PR is:
> m-kuhn wants to merge 28 commits into microsoft:master from opengisch:fix-android-ndk-clang

Now in your vcpkg clone, given `opengisch:fix-android-ndk-clang` you can do:
~~~
git fetch https://github.com/opengisch/vcpkg fix-android-ndk-clang
git checkout -b opengisch-android-pr FETCH_HEAD
~~~

**Comment by Thomas1664** [comment]
@PhoebeHui I think this issue can be closed given that Android triplets are built in CI
